### DISCUSSION TITLE
aluds.go

### DISCUSSION THREAD

**Discussion by Yzgll** [sarchlab/mgpusim#68] [Q&A]
There are a lot of things I don't understand when I modify my instructions based on what mgpusim has implemented. 
First of all, `dst0offset `and `data0offset` when do they equal `uint(8 + 64*4), uint(8 + 64*4 + 256*4), uint(8 + 64*4 + 256*4*2)` and how do we get this?
 Second, when the instruction is described as `DS[ADDR+offset0*4] = D0; DS[ADDR+ Offset1*4]` In this case, `addr0 := layout.ADDR[i] + Int.offset0 *4` and `addr1 := layout.ADDR[i] + Int.offset1 *4` I can understand, But in the case of `DS[A] = D0 `and `R = DS[A]`, can addr only be `addr0 := layout.ADDR[i] + st.offset0`? 
 Third, the `runDSREAD2B32` function needs to write data to two registers why only `dst0offset` and not `dst1offse`?
  Fourth, how can we ensure that `DS[A] = D0[7:0]` reaches the lower eight? `&0xFF`?
  Fifth, `go test `can't run the test file, showing something undecided, but the file is fine.

**Comment by syifan** [comment]
1. I guess you are asking how the `8+64*4` is calculated? It is the data located in the scratchpad (models the input/output registers of ALUs). You can check the scratchpad layout [here](https://github.com/sarchlab/mgpusim/blob/v3/emu/scratchpad.go#L191). 
2. Can you check the [ISA documentation](https://www.amd.com/content/dam/amd/en/documents/radeon-tech-docs/instruction-set-architectures/gcn3-instruction-set-architecture.pdf)? I do not remember the details. 
3. It looks like a bug in our disassembler? The ROCm disassembler generates both offsets. Can you confirm?
4. I do not understand this question. Can you give me more context?
5. What about just run `go test` without a file specified?

**Reply by Yzgll** [reply]
1.OK
2.I have been looking at the ISA documentation, maybe I expressed it inaccurately enough, tomorrow I will write a detailed one.
3.Do you mean that the aluds.go file was generated by ROCm disassembler and not written?
4.I mean, for example, this instruction, DS_WRITE_B8: DS[A] = D0[7:0]; byte write. It requires the 0-7 bits of D0 to be written to DS[A], so how do I ensure that the data I fetch is the 0-7 bits? By performing a bitwise AND operation with 0xFF?
5.I haven't tried this, but I'll try.

**Reply by syifan** [reply]
3. You can use `clang-ocl` to generate an assembly file with a `-S` argument. Then you can compare if the MGPUSim disassembler can generate the same instructions. We should always fix the disassembler before fixing the emulator. 
4. Unless the document is saying something else, yes, masking all the other bits to 0 should be done.

**Reply by Yzgll** [reply]
I know the assembly file, but what is MGPUSim disassembler? How do I get it?

**Reply by syifan** [reply]
Here it is. https://github.com/sarchlab/mgpusim/tree/v3/insts/gcn3disassembler

**Reply by Yzgll** [reply]
Today, the two instructions I modified, `runDSWRITEB8` and `runDSREADI8`, finally passed the tests. However, after running `./aes -timing -verify`, I still encountered a "`Mismatch at position`" error. I checked the `kernels.asm` and the disassembly output and found two issues:

1.The disassembly result at the `ds_write_b8 `instruction differs from the `kernels.asm` file.

2.A panic occurred at the `ds_read_i8 `instruction in the disassembly, showing the error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4ec892]
```
This caused the disassembly to stop running, resulting in missing subsequent instructions compared to `kernels.asm`.

What could be the cause of these two issues? Could it be a problem with the ROCm version? I am using version 3.8. Would you like to see the two instructions and the tests I added?
                                                              **Figure 1: All instructions passed the tests.**
![1](https://github.com/sarchlab/mgpusim/assets/165646862/8cd5866a-bf77-497a-af2c-856d018f0d7c)
**Figure 2: The `ds_write_b8` instruction in kernels.asm.**
![2](https://github.com/sarchlab/mgpusim/assets/165646862/d5ffa668-9034-458d-991a-7ed69d4d7cf6)
**Figure 3: The `ds_write_b8` instruction in the disassembly.**
![3](https://github.com/sarchlab/mgpusim/assets/165646862/4e82be63-5017-4474-b462-c4c9e021482f)
**Figure 4: The `ds_read_i8` instruction in kernels.asm.**
![4](https://github.com/sarchlab/mgpusim/assets/165646862/1daadbab-d31f-41f6-9650-5adf0357c033)
**Figure 5: This is where the `ds_read_i8` instruction should be in the disassembly, but a panic occurred instead.**
![5](https://github.com/sarchlab/mgpusim/assets/165646862/ec2565a3-fa52-4b4b-996d-53f301440ae7)

**Reply by syifan** [reply]
Can you help with debugging the disassembly problem? It should be relatively easy to solve.

**Reply by Yzgll** [reply]
What specifically do you mean by the disassembly problem?

**Reply by syifan** [reply]
Since you experience a panic when you disassemble the program with MGPUSim, can you fix MGPUSim so that it can at least disassemble the whole file?

**Reply by Yzgll** [reply]
Sure, I'll do my best. But how should we resolve the issue where the instructions in the disassembly differ from those in kernels.asm?

**Reply by syifan** [reply]
It is the same debugging process. You will need to identify the code that generates the wrong instruction and find out what the problem is.

**Reply by Yzgll** [reply]
> Since you experience a panic when you disassemble the program with MGPUSim, can you fix MGPUSim so that it can at least disassemble the whole file?

I found the cause; it seems to be my error rather than MGPUSim's. After I modified this https://github.com/sarchlab/mgpusim/blob/v3/insts/inst.go#L299, it could disassemble the entire file. The panic at `ds_read_i8` is gone, and `ds_write_b8` has also changed.

Figure 1："v1" was supposed to be the destination register, but it has now become the source register, while the destination register is nil.
![5](https://github.com/sarchlab/mgpusim/assets/165646862/aff4e9ba-88f7-4924-85bc-c02fe2a48151)

Figure 2："The destination register is nil."
![6](https://github.com/sarchlab/mgpusim/assets/165646862/3a402b91-a0d8-4784-b727-5e0cd81d6b97)

Here is my analysis:
I believe the root cause is the incorrect implementation of the instruction in the `aluds.go` file, with the error likely in the calculation of `8 + 64 * 4`. To be honest, I also reviewed the scratchpad layout, but I still don't understand how this calculation is derived, so I assigned the value directly without calculating it. It happened to pass the test. Therefore, I think this is the key issue.
Do you think my analysis is reasonable? Or could there be other reasons?

**Reply by syifan** [reply]
I do not think the error has anything to do with `aluds.go`. The `aluds.go` code is for the emulator. The disassembler does not need to call the emulator code.

**Reply by Yzgll** [reply]
Yes, you are right. I also thought about it for a long time yesterday. The disassembled file is `kernels.hsaco`, which is compiled from `kernels.cl `and has nothing to do with `aluds.go`.
I found the cause of the panic. When I disassembled, I printed out the relevant instruction information and found that `I.Dst = nil`. However, at https://github.com/sarchlab/mgpusim/blob/v3/insts/inst.go#L303, `I.Dst.String()` was called. Therefore, `ds_read_i8 `caused the panic, while `ds_write_b8` (opcode 30) did not call this.Why is their Dst empty?
![1](https://github.com/sarchlab/mgpusim/assets/165646862/1905e7e3-1736-4d0e-a9a0-457daf8767ae)
![2](https://github.com/sarchlab/mgpusim/assets/165646862/3fd957a7-9b14-4e23-bd45-7008bceca0d3)
![9](https://github.com/sarchlab/mgpusim/assets/165646862/6ae1241a-460a-4fca-bc2b-145a6551da9b)
Another issue I've discovered is that in the disassembly, the `Dst` and `Addr` of `ds_write_b8 `seem to be reversed compared to `kernels.asm`
